src/Swarm/Game/Step.hs

```diff
@@ -968,6 +968,14 @@ execConst c vs s k = do
 
         return $ Out res s k
       _ -> badConst
+    Knows -> case vs of
+      [VString name] -> do
+        inv <- use robotInventory
+        let knows = case E.lookupByName name inv of
```

Comment:
```suggestion
        ins <- use installedDevices
        let allKnown = inv `E.union` ins
        let knows = case E.lookupByName name allKnown of
```
The same thing is needed in `place`, `give` and `install`. #410 

My take is that it would be best to install via a lens that adds to inventory a zero so that an invariant is maintained.

But that should not hold back this PR, with this change it should be good to go. :+1:

Reply:
> My take is that it would be best to install via a lens that adds to inventory a zero so that an invariant is maintained.

Ah, that's a good idea.